acme_diags/driver/utils/diurnal_cycle.py

```diff
@@ -0,0 +1,152 @@
+import numpy 
+import numpy.ma as ma
+import cdms2
+import MV2
+
+
+def composite_diurnal_cycle(var, season):
+    """
+    Compute the composite diurnal cycle for var for the given season.
+    Return mean + amplitudes and times-of-maximum of the first Fourier harmonic component as three transient variables.  
+    """
+    season_idx = {
+        '01': [1,0,0,0,0,0,0,0,0,0,0,0],
+        '02': [0,1,0,0,0,0,0,0,0,0,0,0],
+        '03': [0,0,1,0,0,0,0,0,0,0,0,0],
+        '04': [0,0,0,1,0,0,0,0,0,0,0,0],
+        '05': [0,0,0,0,1,0,0,0,0,0,0,0],
+        '06': [0,0,0,0,0,1,0,0,0,0,0,0],
+        '07': [0,0,0,0,0,0,1,0,0,0,0,0],
+        '08': [0,0,0,0,0,0,0,1,0,0,0,0],
+        '09': [0,0,0,0,0,0,0,0,1,0,0,0],
+        '10': [0,0,0,0,0,0,0,0,0,1,0,0],
+        '11': [0,0,0,0,0,0,0,0,0,0,1,0],
+        '12': [0,0,0,0,0,0,0,0,0,0,0,1],
+        'DJF':[1,1,0,0,0,0,0,0,0,0,0,1],
+        'MAM':[0,0,1,1,1,0,0,0,0,0,0,0],
+        'JJA':[0,0,0,0,0,1,1,1,0,0,0,0],
+        'SON':[0,0,0,0,0,0,0,0,1,1,1,0],
+        'ANN':[1,1,1,1,1,1,1,1,1,1,1,1],
+    }
+
+    # Redefine time to be in the middle of the time interval
+    var_time = var.getTime()
+    if var_time is None:
+        # Climo cannot be ran on this variable.
+        return var
+
+#    tbounds = var_time.getBounds()
+#    var_time[:] = 0.5*(tbounds[:,0]+tbounds[:,1]) #time bounds for h1-h4 are problematic
+    var_time_absolute = var_time.asComponentTime()
+    time_freq = int(24/(var_time_absolute[1].hour - var_time_absolute[0].hour)) #This only valid for time interval >= 1hour
+    start_time = var_time_absolute[0].hour
+    print('start_time',var_time_absolute[0],var_time_absolute[0].hour)
+    print('var_time_freq',time_freq)
+
+    # Convert to masked array
+    v = var.asma()
+
+    # select specified seasons:
+    if season == 'ANNUALCYCLE':  #Not supported yet!
+        cycle = ['01','02','03','04','05','06','07','08','09','10','11','12']
+    elif season == 'SEASONALCYCLE': #Not supported yet!
+        cycle = ['DJF','MAM','JJA','SON']
+    else:
+        cycle = [season]
+
+    ncycle = len(cycle)
+    for n in range(ncycle):
+        idx = numpy.array( [ season_idx[cycle[n]][var_time_absolute[i].month-1]
+                          for i in range(len(var_time_absolute)) ], dtype=numpy.int).nonzero()
+
+        var_season = ma.zeros([ncycle]+[len(idx[0])]+list(numpy.shape(v))[1:])
+        var_season[n,] = v[idx]
+    var_daily = numpy.reshape(var_season,(ncycle,int(var_season.shape[1]/time_freq),time_freq,var_season.shape[2],var_season.shape[3]))
```

Comment:
Did you check that the `reshape` returns the proper results? For instance, in Streamflow Diags, there was a call to `reshape` creating rows of `[January of year 1, February of year 1,...]` rather than `[January of year 1, January of year 2,...]`.